[BUG] Nightly CI of non-UTC time zone reports 'year 0 is out of range' error

**Describe the bug**
In non-UTC time zone CI, it reports:
```
[2023-12-04T03:24:48.506Z] =========================== short test summary info ============================

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] = 16 failed, 19637 passed, 2021 skipped, 371 xfailed, 403 xpassed, 915 warnings in 5621.07s (1:33:41) =
```
It's because string regexp produce not-supported timestamp string.

**Expected behavior**
It's blocking branch-23.12, so fix seed can fix fast, then handle it later properly.

**Environment details (please complete the following information)**

**Additional context**

--- a/integration_tests/src/main/python/cast_test.py
+++ b/integration_tests/src/main/python/cast_test.py
@@ -61,13 +61,11 @@ def test_cast_nested(data_gen, to_type):
     assert_gpu_and_cpu_are_equal_collect(
             lambda spark : unary_op_df(spark, data_gen).select(f.col('a').cast(to_type)))
 
-date_after_1_2_1 = '(0{0,3}1-(0?[2-9]|[1-3][0-9]))|(([0-9]{0,3}[2-9]|[1-9][0-9]{0,2}[01])-[0-3]?[0-9])-[0-5]?[0-9]'
-
 def test_cast_string_date_valid_format():
     # In Spark 3.2.0+ the valid format changed, and we cannot support all of the format.
     # This provides values that are valid in all of those formats.
     assert_gpu_and_cpu_are_equal_collect(
-            lambda spark : unary_op_df(spark, StringGen(date_after_1_2_1)).select(f.col('a').cast(DateType())),
+            lambda spark : unary_op_df(spark, StringGen(date_start_1_2_1)).select(f.col('a').cast(DateType())),
             conf = {'spark.rapids.sql.hasExtendedYearValues': 'false'})
 
 invalid_values_string_to_date = ['200', ' 1970A', '1970 A', '1970T',  # not conform to "yyyy" after trim
@@ -148,9 +146,9 @@ def test_cast_string_date_non_ansi():
         lambda spark: spark.createDataFrame(data_rows, "a string").select(f.col('a').cast(DateType())),
         conf={'spark.rapids.sql.hasExtendedYearValues': 'false'})
 
-@pytest.mark.parametrize('data_gen', [StringGen(date_after_1_2_1),
-                                      StringGen(date_after_1_2_1 + '[ |T][0-3][0-9]:[0-6][0-9]:[0-6][0-9]'),
-                                      StringGen(date_after_1_2_1 + '[ |T][0-3][0-9]:[0-6][0-9]:[0-6][0-9]\.[0-9]{0,6}Z?')
+@pytest.mark.parametrize('data_gen', [StringGen(date_start_1_2_1),
+                                      StringGen(date_start_1_2_1 + '[ |T][0-3][0-9]:[0-6][0-9]:[0-6][0-9]'),
+                                      StringGen(date_start_1_2_1 + '[ |T][0-3][0-9]:[0-6][0-9]:[0-6][0-9]\.[0-9]{0,6}Z?')
                                       ],
                         ids=idfn)
 @allow_non_gpu(*non_utc_allow)

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -779,6 +779,7 @@ def gen_df_help(data_gen, length, seed_value):
     rand = random.Random(seed_value)
     data_gen.start(rand)
     data = [data_gen.gen() for index in range(0, length)]
+    print(data)
     return data
 
 def gen_df(spark, data_gen, length=2048, seed=None, num_slices=None):
@@ -1201,3 +1202,9 @@ def get_25_partitions_df(spark):
 non_utc_allow = ['ProjectExec', 'FilterExec', 'FileSourceScanExec', 'BatchScanExec', 'CollectLimitExec',
                  'DeserializeToObjectExec', 'DataWritingCommandExec', 'WriteFilesExec', 'ShuffleExchangeExec',
                  'ExecutedCommandExec'] if is_not_utc() else []
+
+# date related regexps for generating date strings within python's range limits
+
+date_start_1_2_1 = '(0{0,3}1-(0?[2-9]|[1-3][0-9]))|(([0-9]{0,3}[2-9]|[1-9][0-9]{0,2}[01])-[0-3]?[0-9])-[0-5]?[0-9]'
+
+yyyy_start_0003 = '([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])'

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -779,7 +779,6 @@ def gen_df_help(data_gen, length, seed_value):
     rand = random.Random(seed_value)
     data_gen.start(rand)
     data = [data_gen.gen() for index in range(0, length)]
-    print(data)
     return data
 
 def gen_df(spark, data_gen, length=2048, seed=None, num_slices=None):

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -1207,5 +1207,5 @@ non_utc_allow = ['ProjectExec', 'FilterExec', 'FileSourceScanExec', 'BatchScanEx
 # regexp to generate date from 0001-02-01, format is yyyy-MM-dd
 date_start_1_2_1 = '(0{0,3}1-(0?[2-9]|[1-3][0-9]))|(([0-9]{0,3}[2-9]|[1-9][0-9]{0,2}[01])-[0-3]?[0-9])-[0-5]?[0-9]'
 
-# regexp to generate year from 0002, forat is yyyy
+# regexp to generate year from 0002, format is yyyy
 yyyy_start_0002 = '([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])'

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -1204,6 +1204,8 @@ non_utc_allow = ['ProjectExec', 'FilterExec', 'FileSourceScanExec', 'BatchScanEx
 
 # date related regexps for generating date strings within python's range limits
 
+# regexp to generate date from 0001-02-01, format is yyyy-MM-dd
 date_start_1_2_1 = '(0{0,3}1-(0?[2-9]|[1-3][0-9]))|(([0-9]{0,3}[2-9]|[1-9][0-9]{0,2}[01])-[0-3]?[0-9])-[0-5]?[0-9]'
 
-yyyy_start_0003 = '([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])'
+# regexp to generate year from 0002, forat is yyyy
+yyyy_start_0002 = '([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])'

--- a/integration_tests/src/main/python/json_test.py
+++ b/integration_tests/src/main/python/json_test.py
@@ -677,11 +677,6 @@ def test_from_json_struct_timestamp(timestamp_gen, timestamp_format, time_parser
               'spark.sql.legacy.timeParserPolicy': time_parser_policy,
               'spark.sql.ansi.enabled': ansi_enabled })
 
-def test_yyyy_start_0003():
-    # 0003 is not a leap year
-    assert_gpu_and_cpu_are_equal_collect(
-        lambda spark : gen_df(spark, StringGen(yyyy_start_0003, nullable=False)))
-
 @allow_non_gpu('ProjectExec')
 @pytest.mark.parametrize('timestamp_gen', ["\"[1-8]{1}[0-9]{3}-[0-3]{1,2}-[0-3]{1,2}T[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}(\\.[0-9]{1,6})?Z?\""])
 @pytest.mark.parametrize('timestamp_format', [

--- a/integration_tests/src/main/python/json_test.py
+++ b/integration_tests/src/main/python/json_test.py
@@ -637,7 +637,7 @@ non_utc_project_allow = ['ProjectExec'] if is_not_utc() else []
     # "yyyy-MM"
     "\"[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[1-8]{1}[0-9]{3}-[0-3]{1,2}[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?\"",
     # "yyyy"
-    "\"[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?" + yyyy_start_0003 + "[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?\"",
+    "\"[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?" + yyyy_start_0002 + "[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?\"",
     # "dd/MM/yyyy"
     "\"[0-9]{2}/[0-9]{2}/[1-8]{1}[0-9]{3}\"",
     # special constant values

--- a/integration_tests/src/main/python/json_test.py
+++ b/integration_tests/src/main/python/json_test.py
@@ -637,7 +637,7 @@ non_utc_project_allow = ['ProjectExec'] if is_not_utc() else []
     # "yyyy-MM"
     "\"[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[1-8]{1}[0-9]{3}-[0-3]{1,2}[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?\"",
     # "yyyy"
-    "\"[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?\"",
+    "\"[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?" + yyyy_start_0003 + "[ \t\xA0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?\"",
     # "dd/MM/yyyy"
     "\"[0-9]{2}/[0-9]{2}/[1-8]{1}[0-9]{3}\"",
     # special constant values
@@ -664,7 +664,6 @@ non_utc_project_allow = ['ProjectExec'] if is_not_utc() else []
     pytest.param("LEGACY", marks=pytest.mark.allow_non_gpu('ProjectExec')),
     "CORRECTED"
 ])
-@datagen_overrides(seed=0, reason='https://github.com/NVIDIA/spark-rapids/issues/9747')
 @pytest.mark.parametrize('ansi_enabled', [ True, False ])
 def test_from_json_struct_timestamp(timestamp_gen, timestamp_format, time_parser_policy, ansi_enabled):
     json_string_gen = StringGen(r'{ "a": ' + timestamp_gen + ' }') \
@@ -677,6 +676,11 @@ def test_from_json_struct_timestamp(timestamp_gen, timestamp_format, time_parser
         conf={"spark.rapids.sql.expression.JsonToStructs": True,
               'spark.sql.legacy.timeParserPolicy': time_parser_policy,
               'spark.sql.ansi.enabled': ansi_enabled })
+
+def test_yyyy_start_0003():
+    # 0003 is not a leap year
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark : gen_df(spark, StringGen(yyyy_start_0003, nullable=False)))
 
 @allow_non_gpu('ProjectExec')
 @pytest.mark.parametrize('timestamp_gen', ["\"[1-8]{1}[0-9]{3}-[0-3]{1,2}-[0-3]{1,2}T[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}(\\.[0-9]{1,6})?Z?\""])

